cf-agent/acl_posix.c

```diff
@@ -63,13 +63,14 @@ static int ACECount(acl_t acl);
 static int PermsetEquals(acl_permset_t first, acl_permset_t second);
 
 
-PromiseResult CheckPosixLinuxACL(EvalContext *ctx, const char *file_path, Acl acl, Attributes a, const Promise *pp)
+PromiseResult CheckPosixLinuxACL(EvalContext *ctx, const char *file_path, Acl acl, const Attributes *a, const Promise *pp)
 {
+    assert(a != NULL);
     PromiseResult result = PROMISE_RESULT_NOOP;
 
     if (!CheckPosixLinuxAccessACEs(ctx, acl.acl_entries, acl.acl_method, file_path, a, pp, &result))
     {
-        cfPS(ctx, LOG_LEVEL_ERR, PROMISE_RESULT_FAIL, pp, &a, "Failed checking access ACL on %s", file_path);
+        cfPS(ctx, LOG_LEVEL_ERR, PROMISE_RESULT_FAIL, pp, a, "Failed checking access ACL on %s", file_path);
```

Comment:
So cfPS() has an assert(attr != NULL) but also a dereference ```attr->classes```. Should we be checking if attr != NULL "for real" and not just in a debug assert()? Previously this was a bit more safe because the ```Attributes a``` is a struct which cannot be NULL right?

Reply:
> So cfPS() has an assert(attr != NULL) but also a dereference `attr->classes`. Should we be checking if attr != NULL "for real" and not just in a debug assert()?

Probably not a bad idea, it would be a shame if a logging function caused a *segfault*.

Reply:
@craigcomstock The general idea is that this function should never be called with a NULL pointer. Callers should always make sure that they are not calling with a NULL pointer. The assert communicates that it is invalid to call this function with a NULL pointer. The same applies to the other pointers `pp` and `ctx`, as far as I can tell, but they are not so clearly communicated with `assert`. I really wish that C had a `non-null` pointer type, similar to references in C++.

When creating a lot of functions which take pointers it doesn't make sense to recheck if they are NULL in the beginning of every function. This pattern is pretty common (remember the openssl functions we looked at a couple of weeks ago).

_With all of that being said_, in this specific case, I think it's fine to add 1 (or 3) `if`'s to the beginning of `cfPS`, just in case.

Reply:
Cool. Yeah, was wondering what a good pattern was. It makes sense to me to have the asserts and if you are getting good test coverage during development then you are pretty well off. _if_ you have good test coverage of course ;)